### PR TITLE
Allow auth handler plugins to match URL via glob-like pattern in channel_settings

### PR DESCRIPTION
**Pull request by mattkram**
### Description

Closes: #13778

Implements the ability to specify a glob-like URL pattern in `channel_settings.channel` instead of requiring an exact URL match. It should be noted that if a pattern is used the URL scheme must be specified and match exactly.

1. Check for exact match (existing behavior).
2. If that fails, we parse the channel URL and the target settings URL.
3. If the HTTP schema is missing from the pattern or not identical to the channel URL, we do not match
4. Otherwise, we check for a glob-like pattern match using fnmatch.

### Example usage

Given the following `.condarc`, both specified channels would match the listed auth handler plugin:

```yaml
channels:
  - https://repo.some-hostname.com/some-channel
  - https://repo.some-hostname.com/some-other-channel

channel_settings:
  - channel: https://repo.some-hostname.com/*
    auth: some-auth-handler-plugin
```

### Testing notes

I have added an additional test which covers cases of specifying an exact URL match and a pattern match, as well as negative matches. For this I have used `pytest.mark.parametrize` with named parametrizations.

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [x] Add / update outdated documentation?

### CODE REVIEW

**Review by mcg1969 — DISMISSED** [review]

**Review by intentionally-left-nil — APPROVED** [review]

**Review by jezdez — CHANGES_REQUESTED** [review]
This needs to use a full pattern matching technique instead of special-casing the suffix given the non-trivial effort to fix the inevitable request to expand the scheme to path wildcards, subdomain wildcards etc in the future. I would suggest to use Python's fnmatch, assuming we can do a cross-check if there are security or performance concerns. We could pre-compile the patterns and cache them, if they aren't already in the stdlib variant.

To reduce the risk of downgrade attacks, I would suggest to not allow wildcards for the scheme portion of the channel URL (you have to spell out `http://` if you must).

File: tests/gateways/test_connection.py [hunk 1]
```diff
@@ -187,6 +187,54 @@ def test_get_session_with_channel_settings(mocker):
...
+    assert type(session_obj) is CondaSession
+
+    # For session objects with a custom auth handler it will not be set to CondaHttpAuth
+    assert type(session_obj.auth) is not CondaHttpAuth
+
+    # Make sure we tried to retrieve our auth handler in this function
+    assert (
+        mocker.call("dummy_one")
+        in mock_context.plugin_manager.get_auth_handler.mock_calls
+    )
```
[the remark this answers is not in the document]
Comment (mattkram) [reply tests/gateways/test_connection.py:235]: 🤦 Definitely forgot to finish my refactor here. Thanks for the catch!

**Comment by intentionally-left-nil** [comment]
@jezdez I think the `*` approach doesn't prevent a deeper regex pattern in the future. This is a nice balance of solving an immediate need with a specific usecase. But that's my 2 cents

**Comment by dholth** [comment]
Could we re-use some kind of existing "browser cookie rules" engine to handle this and not regex?

**Comment by mattkram** [comment]
@jezdez thank you for the feedback! I have implemented your recommendation. At a high level, we follow the following logic:

1. Check for exact match (existing behavior).
2. If that fails, we parse the channel URL and the target settings URL.
3. If the HTTP schema is missing from the pattern or not identical to the channel URL, we do not match
4. Otherwise, we check for a glob-like pattern match using `fnmatch`.

Please let me know if you think the test is over-parametrized. I'm trying to balance code duplication vs simplicity and I may be crossing a barrier where this should be two different tests.

**Review by travishathaway — APPROVED** [review]

File: conda/gateways/connection/session.py [hunk 2]
```diff
@@ -89,7 +90,24 @@ def get_session(url: str):
...
+            continue
+
+        # If we don't have an exact match, we attempt to match a URL pattern
+        parsed_url = urlparse(url)
+        parsed_setting = urlparse(channel)
+
+        # We require that the schemes must be identical to prevent downgrade attacks.
+        # This includes the case of a scheme-less pattern like "*", which is not allowed.
+        if parsed_setting.scheme != parsed_url.scheme:
+            continue
```
Comment (jezdez) [inline conda/gateways/connection/session.py:106]: Nice!

**Review by jezdez — APPROVED** [review]

**Pull request merged**